cli/compose/loader/loader.go

```diff
@@ -412,32 +412,48 @@ func resolveVolumePaths(volumes []types.ServiceVolumeConfig, workingDir string,
 			return errors.New(`invalid mount config for type "bind": field Source must not be empty`)
 		}
 
-		filePath := expandUser(volume.Source, lookupEnv)
-		// Check for a Unix absolute path first, to handle a Windows client
-		// with a Unix daemon. This handles a Windows client connecting to a
-		// Unix daemon. Note that this is not required for Docker for Windows
-		// when specifying a local Windows path, because Docker for Windows
-		// translates the Windows path into a valid path within the VM.
-		if !path.IsAbs(filePath) {
-			filePath = absPath(workingDir, filePath)
+		filePath, err := transformFilePath(volume.Source, workingDir, lookupEnv)
+		if err != nil {
+			return err
 		}
+
 		volume.Source = filePath
 		volumes[i] = volume
 	}
 	return nil
 }
 
-// TODO: make this more robust
-func expandUser(path string, lookupEnv template.Mapping) string {
-	if strings.HasPrefix(path, "~") {
+func transformFilePath(origin string, workingDir string, lookupEnv template.Mapping) (string, error) {
+	path, err := crosspath.ParsePathWithDefaults(origin)
+	if err != nil {
+		return "", err
+	}
+	switch path.Kind() {
+	case crosspath.Relative:
+		basePath, err := crosspath.ParsePathWithDefaults(workingDir)
+		if err != nil {
+			return "", err
+		}
+		path, err = basePath.Join(path)
+		if err != nil {
+			return "", err
+		}
+	case crosspath.HomeRooted:
 		home, ok := lookupEnv("HOME")
```

Comment:
Instead of patching `$HOME`, should we just always `homedir.Get()`? e.g. on Windows, the variable would be `% USERPROFILE%` https://github.com/moby/moby/blob/4f0d95fa6ee7f865597c03b9e63702cdcb0f7067/pkg/homedir/homedir_windows.go#L7-L11

Reply:
I'll change that (however, it makes for a bigger code change than the original one). I agree it seems a bit better